File: Python/Minimum Window Substring.py
class Solution:
    # @return a string
    def minWindow(self, S, T):
		count1 = {};count2={}
		for char in T:
			if char not in count1: count1[char]=1
			else: count1[char]+=1
		for char in T:
			if char not in count2: count2[char]=1
			else: count2[char]+=1
		start = 0
		count = len(T)
		minSize = 100000
		minStart = 0
		for end in range(len(S)):
			if S[end] in count2 and count2[S[end]]>0:#在第一次找到substring的时候，这个substring包含所有在T里的字母，，那么count1[i]<=0,在下一次遇到任何一个字母时count==0的条件会自动满足，
														#无需进入count1[S[end]]>=0的判断
				count1[S[end]] = count1[S[end]] - 1
				if count1[S[end]]>=0:#count1 记录在currentWindow里多余的字母的个数= abs(count1[i])
					count = count - 1
				if count == 0:
					while True:
						if S[start] in count2 and count2[S[start]]>0:
							if count1[S[start]] < 0:
								count1[S[start]] = count1[S[start]]+1
							else:
								break
						start = start + 1
					if minSize>end-start+1:
						minSize = end - start + 1
						minStart = start
		if minSize == 100000:return ""
		else: return S[minStart:minStart+minSize]
class Solution:
    # @return a string
    def minWindow(self, S, T):
        count1={}; count2={}
        for char in T:
            if char not in count1: count1[char]=1
            else: count1[char]+=1
        for char in T:
            if char not in count2: count2[char]=1
            else: count2[char]+=1
        count=len(T)
        start=0; minSize=100000; minStart=0
        for end in range(len(S)):
            if S[end] in count2 and count2[S[end]]>0:
                count1[S[end]]-=1
                if count1[S[end]]>=0:
                    count-=1
            if count==0:
                while True:
                    if S[start] in count2 and count2[S[start]]>0:
                        if count1[S[start]]<0:
                            count1[S[start]]+=1
                        else:
                            break
                    start+=1
                if minSize>end-start+1:
                    minSize=end-start+1; minStart=start
        if minSize==100000: return ''
        else:
            return S[minStart:minStart+minSize]

File: Java/Minimum Window Substring.java
class Minimum Window Substring{
	public String minWindow(String S, String T) {
        Map<Character,Integer> map1 = new HashMap<Character,Integer>();
        Map<Character,Integer> map2 = new HashMap<Character,Integer>();
        for(int i=0;i<T.length();i++){
            if(map1.containsKey(T.charAt(i)))
                map1.put(T.charAt(i),1+map1.get(T.charAt(i)));
            else
                map1.put(T.charAt(i),1);
            if(map2.containsKey(T.charAt(i)))
                map2.put(T.charAt(i),1+map2.get(T.charAt(i)));
            else
                map2.put(T.charAt(i),1);
        }
        int countOfT = T.length();
        int end = 0;
        int start = 0;
        int minStart = 0;
        int minSize = Integer.MAX_VALUE;
        //map1 used to calc the redudent characters abs()
        while(end < S.length()){
            if(map2.containsKey(S.charAt(end))){
                if(map1.get(S.charAt(end)) > 0)
                    countOfT--;
                map1.put(S.charAt(end),map1.get(S.charAt(end))-1);
                //in current substring of s1 all s2 charcters has occured
                if(countOfT == 0){
                    while(start <= end){
                        if(map2.containsKey(S.charAt(start))){
                        	//if the redundent characters in map1 still bigger than 0
                            if(map1.get(S.charAt(start)) < 0)
                                map1.put(S.charAt(start),map1.get(S.charAt(start))+1);
                            //if there is no redundency
                            else
                                break;
                        }
                        start++;
                    }
                    
                    if (minSize > end - start + 1){
                        minStart = start;
                        minSize = end - start + 1;
                    }
                }
            }
            end++;
        }
        if (minSize == Integer.MAX_VALUE)
            return "";
        else{
            return S.substring(minStart,minStart + minSize);
        }
            
    }
}